src/magelang/passes/check_undefined.py

```diff
@@ -0,0 +1,18 @@
+
+
+from magelang.ast import Grammar, RefExpr, Expr
+def check_undefined(grammar: Grammar) -> Grammar:
+    def traverse(expr: Expr) -> None:
+        if isinstance(expr, RefExpr):
+            if not any(rule.name == expr.name for rule in grammar.rules):
```

Comment:
`grammar.lookup` is more performant in this case, because it uses a hash table. It needs some adjustments though. Will do this in the next commit.